localstack-core/localstack/services/stepfunctions/asl/component/state/state_execution/execute_state.py

```diff
@@ -213,7 +216,7 @@ def _eval_state(self, env: Environment) -> None:
         env.context_object_manager.context_object["State"]["RetryCount"] = 0
 
         # Attempt to evaluate the state's logic through until it's successful, caught, or retries have run out.
-        while True:
+        while env.is_running():
```

Comment:
To clarify, is it deseriable to exit early when the program is in an error state?

```python
    def is_running(self) -> bool:
        return isinstance(self._program_state, ProgramRunning)
```

I'm wondering if there's a possibility of an early exit here since is there not a possibility of the program/env going into an error state (meaning `_program_state != ProgramRunning`) and this loop breaking prematurely since the condition is false?

Reply:
Yes this is the correct behaviour in this case.
The program state is changed only once after startup, and signals the program ending, terminal error, or stop request. All child threads should quite as soon as possible to allow the main thread (of the state machine execution) to return.